airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/comparator/AdvancedTestDataComparator.java

```diff
@@ -0,0 +1,191 @@
+        assertSameValue(expectedValue, actualValue);
+      }
+    }
+  }
+
+  private boolean areBothEmpty(final JsonNode expectedData, final JsonNode actualData) {
+    if (expectedData.size() == 0 && actualData.size() == 0) {
+      return true;
+    } else if (expectedData.size() <= 1 && actualData.size() <= 1) {
+      boolean isExpectedEmpty = (expectedData.size() == 0 || expectedData.iterator().next().asText().isEmpty());
+      return isExpectedEmpty && (actualData.size() == 0 || actualData.iterator().next().asText().isEmpty());
+    } else {
+      return false;
+    }
+  }
+
+  // Allows subclasses to implement custom comparison asserts
+  protected void assertSameValue(final JsonNode expectedValue, final JsonNode actualValue) {
+    LOGGER.info("assertSameValue : {} vs {}", expectedValue, actualValue);
+
+    assertTrue(compareJsonNodes(expectedValue, actualValue), "Expected value " + expectedValue + " vs Actual value " + actualValue);
+  }
+
+  protected boolean compareJsonNodes(final JsonNode expectedValue, final JsonNode actualValue) {
+    if (expectedValue == null || actualValue == null)
+      return expectedValue == null && actualValue == null;
+    else if (expectedValue.isNumber() || expectedValue.isDouble() || expectedValue.isFloat())
+      return compareNumericValues(expectedValue.asText(), actualValue.asText());
+    else if (expectedValue.isBoolean())
+      return compareBooleanValues(expectedValue.asText(), actualValue.asText());
+    else if (isDateTimeWithTzValue(expectedValue.asText()))
+      return compareDateTimeWithTzValues(expectedValue.asText(), actualValue.asText());
+    else if (isDateTimeValue(expectedValue.asText()))
+      return compareDateTimeValues(expectedValue.asText(), actualValue.asText());
+    else if (isDateValue(expectedValue.asText()))
+      return compareDateValues(expectedValue.asText(), actualValue.asText());
+    else if (expectedValue.isArray() && actualValue.isArray())
+      return compareArrays(expectedValue, actualValue);
+    else if (expectedValue.isObject() && actualValue.isObject()) {
+      compareObjects(expectedValue, actualValue);
+      return true;
+    } else
+      return compareString(expectedValue, actualValue);
+  }
+
+  protected boolean compareString(final JsonNode expectedValue, final JsonNode actualValue) {
+    return expectedValue.asText().equals(actualValue.asText());
+  }
+
+  private List<JsonNode> getArrayList(final JsonNode jsonArray) {
+    List<JsonNode> result = new ArrayList<>();
+    jsonArray.elements().forEachRemaining(result::add);
+    return result;
+  }
+
+  protected boolean compareArrays(final JsonNode expectedArray, final JsonNode actualArray) {
+    var expectedList = getArrayList(expectedArray);
+    var actualList = getArrayList(actualArray);
+
+    if (expectedList.size() != actualList.size()) {
+      return false;
+    } else {
+      boolean result = false;
+      for (JsonNode expectedNode : expectedList) {
```

Comment:
is this intentionally comparing each pair of expected/actual nodes? I think it would be better to do something like this: (pseudocode)
```
for each expectedNode:
  index = actualList.firstIndexOf(n -> compareJsonNodes(expectedNode, n))
  assertNotEqual(index, -1)
  actualList.remove(index)
assertEmpty(actualList)
```
I.e. remove all the expectedNodes from the actualList, and fail if either we can't find the expectedNode, or if there were too many actualNodes. (maybe that second assert isn't needed, since we assert that the list sizes are equal?)

Reply:
Good point. In addition there is a bug in the code :)